content/tutorials/spatial_wfs_services/index.Rmd

```diff
@@ -498,6 +504,96 @@ knitr::kable(df)
 
 Which indeed corresponds to the data of the coordinate.
 
-This procedure can also be turned into a function with lat, long and properties of interest as parameters.
+This procedure can also be turned into a function with lat, long and properties of interest as parameters:
+
+```{r}
+#' @title Extract soil properties from Flemish soil map
+#'
+#' @description This function queries the Flemish soil map attributres at a given coordinate,
+#' http://www.geopunt.be/catalogus/datasetfolder/5c129f2d-4498-4bc3-8860-01cb2d513f8f
+#' by using the affiliated WFS service provided by DOV. The user can pick the
+#' properties of interest. A full list of properties is available at
+#' https://www.dov.vlaanderen.be/geoserver/bodemkaart/bodemtypes/wfs?request=DescribeFeatureType
+#' Coordinates should be given as Lambert 72 (epsg 31370). When outside the
+#' Flemish region, an NA value is given for each of the properties.
+#'
+#' @param x_lam The numeric value of the X coordinate as Lambert 72.
+#' @param y_lam The numeric value of the X coordinate as Lambert 72.
+#' @param properties_of_interest A vector or properties, as a subset of these
+#' provided by the webservice. Default Bodemserie, Unibodemtype and Bodemtype.
+#'
+#' @return A data.frame with the properties as column headers.
+#'
+#' @importFrom httr GET
+#' @importFrom jsonlite fromJSON
+#'
+#' @examples
+#' soil_map_data_extraction(173995.67, 212093.44)
+#' soil_map_data_extraction(173995.67, 212093.44,
+#'                          properties_of_interest = c("Eenduidige_legende",
+#'                                                     "Textuurklasse"))
+soil_map_data_extraction <- function(x_lam,
+                                     y_lam,
+                                     properties_of_interest = c("Bodemserie",
+                                                                "Unibodemtype",
+                                                                "Bodemtype")) {
+  if (missing(x_lam) | missing(y_lam)) stop("x_lam and y_lam needed")
```

Comment:
redudant check. The function will return a sensible error without this line of code. The line below is more useful. It requires the object to only contain values listed in the default arguments

```suggestion
  properties_of_interest <- match.arg(properties_of_interest, several.ok = TRUE)
```

Reply:
The `match_arg()` would restrict the choice to Bodemserie, Unibodemtype and Bodemtype. But the soil map has many more fields:

```
 [1] "gid"                                      "id_kaartvlak"                            
 [3] "geom"                                     "Bodemtype"                               
 [5] "Unibodemtype"                             "Bodemserie"                              
 [7] "Beknopte_omschrijving_bodemserie"         "Substraat_legende"                       
 [9] "Gegeneraliseerde_legende"                 "Substraat_code"                          
[11] "Substraat_Vlaanderen"                     "Textuurklasse_code"                      
[13] "Textuurklasse"                            "Drainageklasse_code"                     
[15] "Drainageklasse"                           "Profielontwikkelingsgroep_code"          
[17] "Profielontwikkelingsgroep"                "Fase_code"                               
[19] "Fase"                                     "Variante_van_het_moedermateriaal_code"   
[21] "Variante_van_het_moedermateriaal"         "Variante_van_de_profielontwikkeling_code"
[23] "Variante_van_de_profielontwikkeling"      "Substraat_code_zeepolders"               
[25] "Substraat_zeepolders"                     "Streek_code"                             
[27] "Streek"                                   "Serie_code"                              
[29] "Serie"                                    "Subserie_code"                           
[31] "Subserie"                                 "Type_code"                               
[33] "Type"                                     "Subtype_code"                            
[35] "Subtype"                                  "Eenduidige_legende_titel"                
[37] "Eenduidige_legende"                       "Scan_analoge_bodemkaarblad"              
[39] "Scan_toelichtingsboekje"                  "Scan_bodemkaart5000"                     
[41] "Scan_stippenkaart5000"                    "Type_classificatie"                      
[43] "Bodemtype_per_streek"                     "Kaartbladnr"                             
[45] "codeid"                  
```

Reply:
Maybe fetch the list of the service and test if all user specified options are valid? Another option is to leave the argument blank and when missing set to the default (with a message). A bit like `.groups` in the current implementation of `summarise()`

Reply:
I know but one way that fetches the values , but the downside is that it requires `ows4R` package (which loads slow because it has to load ISO 19139 XML schemas):

``` r
library(ows4R)
#> Loading required package: geometa
#> Loading ISO 19139 XML schemas...
#> Loading ISO 19115 codelists...
library(purrr)
wfs <- "https://www.dov.vlaanderen.be/geoserver/bodemkaart/bodemtypes/wfs"
wfs_client <- WFSClient$new(wfs, 
                            serviceVersion = "1.1.0")
wfs_client$
  getCapabilities()$
  findFeatureTypeByName("bodemkaart:bodemtypes")$
  getDescription() %>%
  map_chr(function(x){x$getName()})
#>  [1] "gid"                                     
#>  [2] "id_kaartvlak"                            
#>  [3] "geom"                                    
#>  [4] "Bodemtype"                               
#>  [5] "Unibodemtype"                            
#>  [6] "Bodemserie"                              
#>  [7] "Beknopte_omschrijving_bodemserie"        
#>  [8] "Substraat_legende"                       
#>  [9] "Gegeneraliseerde_legende"                
#> [10] "Substraat_code"                          
#> [11] "Substraat_Vlaanderen"                    
#> [12] "Textuurklasse_code"                      
#> [13] "Textuurklasse"                           
#> [14] "Drainageklasse_code"                     
#> [15] "Drainageklasse"                          
#> [16] "Profielontwikkelingsgroep_code"          
#> [17] "Profielontwikkelingsgroep"               
#> [18] "Fase_code"                               
#> [19] "Fase"                                    
#> [20] "Variante_van_het_moedermateriaal_code"   
#> [21] "Variante_van_het_moedermateriaal"        
#> [22] "Variante_van_de_profielontwikkeling_code"
#> [23] "Variante_van_de_profielontwikkeling"     
#> [24] "Substraat_code_zeepolders"               
#> [25] "Substraat_zeepolders"                    
#> [26] "Streek_code"                             
#> [27] "Streek"                                  
#> [28] "Serie_code"                              
#> [29] "Serie"                                   
#> [30] "Subserie_code"                           
#> [31] "Subserie"                                
#> [32] "Type_code"                               
#> [33] "Type"                                    
#> [34] "Subtype_code"                            
#> [35] "Subtype"                                 
#> [36] "Eenduidige_legende_titel"                
#> [37] "Eenduidige_legende"                      
#> [38] "Scan_analoge_bodemkaarblad"              
#> [39] "Scan_toelichtingsboekje"                 
#> [40] "Scan_bodemkaart5000"                     
#> [41] "Scan_stippenkaart5000"                   
#> [42] "Type_classificatie"                      
#> [43] "Bodemtype_per_streek"                    
#> [44] "Kaartbladnr"                             
#> [45] "codeid"
```

<sup>Created on 2021-01-25 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> - Session info ---------------------------------------------------------------
#>  setting  value                       
#>  version  R version 4.0.2 (2020-06-22)
#>  os       Windows 7 x64 SP 1          
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  Dutch_Belgium.1252          
#>  ctype    Dutch_Belgium.1252          
#>  tz       Europe/Paris                
#>  date     2021-01-25                  
#> 
#> - Packages -------------------------------------------------------------------
#>  package     * version  date       lib source                         
#>  askpass       1.1      2019-01-13 [1] CRAN (R 4.0.0)                 
#>  assertthat    0.2.1    2019-03-21 [1] CRAN (R 4.0.0)                 
#>  callr         3.5.1    2020-10-13 [1] CRAN (R 4.0.3)                 
#>  class         7.3-17   2020-04-26 [2] CRAN (R 4.0.2)                 
#>  classInt      0.4-3    2020-04-07 [1] CRAN (R 4.0.0)                 
#>  cli           2.2.0    2020-11-20 [1] CRAN (R 4.0.3)                 
#>  crayon        1.3.4    2017-09-16 [1] CRAN (R 4.0.0)                 
#>  curl          4.3      2019-12-02 [1] CRAN (R 4.0.0)                 
#>  DBI           1.1.0    2019-12-15 [1] CRAN (R 4.0.0)                 
#>  desc          1.2.0    2018-05-01 [1] CRAN (R 4.0.0)                 
#>  devtools      2.3.1    2020-07-21 [1] CRAN (R 4.0.2)                 
#>  digest        0.6.27   2020-10-24 [1] CRAN (R 4.0.3)                 
#>  dplyr         1.0.2    2020-08-18 [1] CRAN (R 4.0.2)                 
#>  e1071         1.7-4    2020-10-14 [1] CRAN (R 4.0.3)                 
#>  ellipsis      0.3.1    2020-05-15 [1] CRAN (R 4.0.2)                 
#>  evaluate      0.14     2019-05-28 [1] CRAN (R 4.0.0)                 
#>  fansi         0.4.1    2020-01-08 [1] CRAN (R 4.0.0)                 
#>  fs            1.5.0    2020-07-31 [1] CRAN (R 4.0.2)                 
#>  generics      0.1.0    2020-10-31 [1] CRAN (R 4.0.3)                 
#>  geometa     * 0.6-3    2020-08-04 [1] CRAN (R 4.0.2)                 
#>  glue          1.4.2    2020-08-27 [1] CRAN (R 4.0.2)                 
#>  highr         0.8      2019-03-20 [1] CRAN (R 4.0.0)                 
#>  htmltools     0.5.0    2020-06-16 [1] CRAN (R 4.0.2)                 
#>  httr          1.4.2    2020-07-20 [1] CRAN (R 4.0.2)                 
#>  jsonlite      1.7.2    2020-12-09 [1] CRAN (R 4.0.3)                 
#>  KernSmooth    2.23-17  2020-04-26 [2] CRAN (R 4.0.2)                 
#>  knitr         1.30     2020-09-22 [1] CRAN (R 4.0.3)                 
#>  lattice       0.20-41  2020-04-02 [2] CRAN (R 4.0.2)                 
#>  lifecycle     0.2.0    2020-03-06 [1] CRAN (R 4.0.0)                 
#>  magrittr      2.0.1    2020-11-17 [1] CRAN (R 4.0.3)                 
#>  memoise       1.1.0    2017-04-21 [1] CRAN (R 4.0.0)                 
#>  openssl       1.4.3    2020-09-18 [1] CRAN (R 4.0.3)                 
#>  ows4R       * 0.2      2021-01-13 [1] Github (eblondel/ows4R@75df72b)
#>  pillar        1.4.7    2020-11-20 [1] CRAN (R 4.0.3)                 
#>  pkgbuild      1.2.0    2020-12-15 [1] CRAN (R 4.0.2)                 
#>  pkgconfig     2.0.3    2019-09-22 [1] CRAN (R 4.0.0)                 
#>  pkgload       1.1.0    2020-05-29 [1] CRAN (R 4.0.2)                 
#>  prettyunits   1.1.1    2020-01-24 [1] CRAN (R 4.0.0)                 
#>  processx      3.4.5    2020-11-30 [1] CRAN (R 4.0.3)                 
#>  ps            1.5.0    2020-12-05 [1] CRAN (R 4.0.3)                 
#>  purrr       * 0.3.4    2020-04-17 [1] CRAN (R 4.0.0)                 
#>  R6            2.5.0    2020-10-28 [1] CRAN (R 4.0.3)                 
#>  Rcpp          1.0.5    2020-07-06 [1] CRAN (R 4.0.2)                 
#>  remotes       2.2.0    2020-07-21 [1] CRAN (R 4.0.2)                 
#>  rgdal         1.5-19   2021-01-05 [1] CRAN (R 4.0.3)                 
#>  rlang         0.4.9    2020-11-26 [1] CRAN (R 4.0.3)                 
#>  rmarkdown     2.4      2020-09-30 [1] CRAN (R 4.0.3)                 
#>  rprojroot     2.0.2    2020-11-15 [1] CRAN (R 4.0.3)                 
#>  sessioninfo   1.1.1    2018-11-05 [1] CRAN (R 4.0.0)                 
#>  sf            0.9-7    2021-01-06 [1] CRAN (R 4.0.3)                 
#>  sp            1.4-5    2021-01-10 [1] CRAN (R 4.0.3)                 
#>  stringi       1.5.3    2020-09-09 [1] CRAN (R 4.0.2)                 
#>  stringr       1.4.0    2019-02-10 [1] CRAN (R 4.0.0)                 
#>  testthat      3.0.0    2020-10-31 [1] CRAN (R 4.0.3)                 
#>  tibble        3.0.4    2020-10-12 [1] CRAN (R 4.0.3)                 
#>  tidyselect    1.1.0    2020-05-11 [1] CRAN (R 4.0.2)                 
#>  units         0.6-7    2020-06-13 [1] CRAN (R 4.0.2)                 
#>  usethis       1.6.1    2020-04-29 [1] CRAN (R 4.0.0)                 
#>  vctrs         0.3.5    2020-11-17 [1] CRAN (R 4.0.3)                 
#>  withr         2.4.0    2021-01-16 [1] CRAN (R 4.0.2)                 
#>  xfun          0.19     2020-10-30 [1] CRAN (R 4.0.3)                 
#>  XML           3.99-0.5 2020-07-23 [1] CRAN (R 4.0.2)                 
#>  yaml          2.2.1    2020-02-01 [1] CRAN (R 4.0.0)                 
#> 
#> [1] C:/R/Library
#> [2] C:/Program Files/R/R-4.0.2/library
```

</details>


and

https://www.dov.vlaanderen.be/geoserver/bodemkaart/bodemtypes/wfs?request=DescribeFeatureType

for some to me unknown reason does not work as expected for this wfs.

Unless you know a better alternative, I will probably choose your second suggestion.

Reply:
I recommend to keep the number of dependencies to a minimum. So don't use ows4r. However do mention it in the description of the function. And refer to there from the param information. And see if you get a sensible error when the user supplies a wrong value.

Reply:
see https://github.com/inbo/inborutils/pull/94/commits/6567d1018de6f58c86d9a62ef55d43742e8151d9 and https://github.com/inbo/inborutils/pull/94/commits/a9b484d54a234a3bc760fc7c2e272014ce2ec30e